### ISSUE TITLE
Question - save (not deploy) to another org

### ISSUE THREAD

**Issue by mikegedelman** [neowit/vim-force.com#61]
Hi there,

First of all thanks a bunch for your work on this. I was pulling my hair out using Eclipse and Sublime until I found out this existed. I've used vim-force pretty much every day for the last year or so. As for my question:

I often do `ApexDeployOne! deploy <test_sandbox>` to move code into my QA sandbox once I'm finished in my personal dev sandbox.

For my dev sandbox, I can just do `ApexSaveOne` and it's pretty fast. Is there a way to allow `ApexSave` to accept another org as an argument like `ApexDeploy` does? Like, is there a technical reason this isn't available? If it's possible, I could take a stab at it myself and create a PR.

The time taken to save vs deploy is about 5s vs 60s, so it would be pretty handy.

Maybe there's a way to do it and I'm just not seeing it in the documentation. 

Thanks,
Mike

**Comment by neowit** [comment]
Hello Mike,

`ApexSave` commands use Tooling API.  

Tooling API works with classes/triggers, etc like with objects. If you have class `MyClass.cls` then in order to update it with Tooling API you have to know its Id first. `vim-force.com` keeps track of objects (e.g. classes) it has created/saved/pulled and "knows" Id of each of those. So when you do `ApexSaveOne` (because we already know class Id) it is just a matter of sending command similar to `Database.update(obj)`.  
In theory it is possible to design a solution to allow `ApexSaveOne` to save `MyClass.cls` in another Org, but the process would be something like this:  
- check if `MyClass.cls` exists there  
- if it does - get Id of `MyClass.cls`  
- finally Save `MyClass.cls` 

Tooling API can not be used to deploy stuff into Production, while Prod is usually one of the primary targets for Deploying to another Org. It is possible to find out if target org is Production or not, but this means yet another API call to run on every deployment, i.e. even more time for the whole operation.

Tooling API has all sorts of problems when you try to Create objects with it (which is a frequent case with Save/Deploy to another Org), especially objects of different types (e.g. Classes & triggers). When I initially tested best approach for `ApexSave` it quickly became apparent that the only reliable way to save a bunch of metadata objects where some objects are new or of mixed types is to use Metadata API.  
When you run `ApexSave` the operation does not always use Tooling API, and in many cases have to fall back to Metadata API. In a Dev environment with no or not much managed stuff installed deployment with Metadata API can be as quick as with Tooling API, so you may not even notice that your `ApexSave` had to fall back to Metadata API.

The bottom line is - `ApexSaveOne` is a potentially viable candidate to allow "save" to another org, but `ApexSave` is not (at least with current state of Tooling API).  

I am not aware of anyone using "`Apex*Deploy*One` to another Org" frequently enough to be worth investing time in implementing "`ApexSaveOne` to another Org". 
if we had  `ApexSaveOne` which can save to another Org while  `ApexSave` can not then there is a room for confusion.
However if there is a strong enough case to implement  "`ApexSaveOne` to another Org" then this may be worth considering.

**Comment by mikegedelman** [comment]
Thanks for your quick reply.  That makes sense. I might just write a script to do this one-off since I end up having to do it a lot.. but it sounds like it's probably not a major use case worth the extra round trips to the server / potentially more confusing commands.

**Comment by neowit** [comment]
I am sure you know that, but just in case - if `vim-force.com` does not support your specific workflow you can use `tooling-force.com.jar` directly. It is designed to be used with scripts and does not have to be called from `vim-force.com`

If you frequently deploy classes/triggers/pages from your dev environment to a Sandbox and found that (when working directly with target sandbox) `ApexSave` in that sandbox is **much** faster than `ApexDeploy` then you could automate this something like this:

Assuming you have two environments
- Dev - source
- Sandbox “Staging” - destination

Write a script to:
- run `tooling-force.com.jar` `refresh` action against “Staging”
- copy files you want to deploy to “Staging” sandbox local `src` folder
- run `tooling-force.com.jar` `saveModified` action against “Staging”

If you know that you will be deploying files which are already present in your local copy of "Staging" then `refresh` step is optional, omitting it may reduce deployment time.

**Comment by mikegedelman** [comment]
Awesome, thanks a bunch. Is there a way for me to pass in the list of files through stdin? Or does the list have to be in a file and use --specificFiles='/path/to/files.txt'

**Comment by neowit** [comment]
Sorry not `std[in/out]` support. You have to use `--specificFiles='/path/to/files.txt'` to provide files and read results from file specified by `--responseFilePath`
